### DISCUSSION TITLE
Could we use multiple languages in the same document?

### DISCUSSION THREAD

**Discussion by TS-CUBED** [quarto-dev/quarto-cli#4232] [Q&A] [answered]
At the moment, Quarto doesn't allow mixing languages.

Is this something that you think about supporting in the future?

Since it seems to use an `ipynb` as an interim format, that may be the limiting factor. But if I understand correctly, it's just using the jupyter notebook to create a Markdown document, so using different kernels for the different languages should be possible, and then splicing the outputs back into the common Markdown.

**Comment by cscheid** [comment, marked as answer]
> Since it seems to use an ipynb as an interim format, that may be the limiting factor. 

You've pinpointed the central issue.

> But if I understand correctly, it's just using the jupyter notebook to create a Markdown document, so using different kernels for the different languages should be possible, and then splicing the outputs back into the common Markdown.

That's easier said than done in the generality that we'd be comfortable with in order to make it a quarto feature. To give a few examples:

- how do we resolve the order in which the kernels are selected? What happens when they overlap in the cells that execute?
- how do we actually schedule the execution? Repeated calls into nbclient would require a substantial reworking of our execution codepath.
- how do we teach users about this? We already have a large source of confusion in the case where both python and r code cells are present in the `knitr` engine (which uses the R `reticulate` package).
- will this actually work like users expect? If users expect cells to communicate with each other, then this will need support from the kernels (which by and large won't expect to be executing alongside other kernels)

It's a natural idea, but we currently don't think we can execute on it at the level we prefer.

**Reply by TS-CUBED** [reply]
It's been solved for a while in org-babel (org-mode).

Interaction between kernels is not direct in  org-babel as well (as you said, it would need support at kernel level). But one can use variables that exist outside the kernels (via an exports/results mechanism) to move values between the different kernels.

This latter would be a need thing to have anyway, so one could calculate a result in a code cell and then use it in markdown, e.g.:

`The result is $(result), based on $(parameter).`

or so.

Maybe leave the idea simmering for quarto 2.x

**Reply by cscheid** [reply]
> org-babel

That's not quarto ;)

**Comment by tosh** [comment]
Just wanted to add that for some reason I thought quarto supported multiple languages within the same file (I guess part mis-remembering and mis-understanding that multiple languages != multiple languages in the same file).

**Reply by mcanouil** [reply]
You can have multiple language in the same file, but you cannot have multiple engines.
So if you use an engine supporting several languages, it will work.

**Reply by tosh** [reply]
Ah, I wasn't aware of that. Do you have a recommendation for an engine that supports multiple languages?

**Reply by mcanouil** [reply]
It's highly dependent on the languages you want to use.
So I don't have a recommendation.
I switch between knitr and Jupyter depending on the project/document.

**Reply by floswald** [reply]
for example you could use julia to drive julia, R and python execution via RCall.jl and PyCall.jl - both of which work very well. A drawback is that both your python and R code will be slightly annotated - something like `R"library(ggplot2)"`. So, in order to do actual computation, that works; in order to make it look beautiful, less sure.

**Comment by TS-CUBED** [comment]
I do know (and use) language interaction. In this case I'd like to show and compare languages for teaching purposes. So can't have the annotations.
<https://ts-cubed.github.io/>

________________________________

**Reply by mcanouil** [reply]
One possible solution now, is to use the embed shortcode, see <https://quarto.org/docs/authoring/notebook-embed.html>.

**Comment by TS-CUBED** [comment]
Thanks. Will need to look at that. Does it allow to show the code and output?
<https://ts-cubed.github.io/>

________________________________
